### DISCUSSION TITLE
ActionResult.CONSUME unexpected behavior

### DISCUSSION THREAD

**Discussion by Crystal-Spider** [FabricMC/fabric-api#2131] [Mod Dev Support] [answered]
I'm registering a listener to UseBlockCallback.EVENT.  
My listener needs to do some operations based on whether the block is a fully grown crop or not and if the call is server side, and the server side call happens after the client side.
However if it's in the case where I would do such operations, but it's client side, I can't return PASS otherwise other actions that should not happen will happen client side, which will make stuff go wrong (for example if the player is holding a block and right clicks on a grown crop, the block will look like it got placed for then suddenly returning in the player's hand).
I can't return SUCCESS either because I need to select which hand to play the swinging animation with.
The only thing left is CONSUME which I was expecting to block other actions from happening client side, prevent the animation from playing but still let the server side call to happen.
This does not seem to be the case and I have no idea on how to do differently.

I'm porting from Forge, and on Forge setting the event interaction result to CONSUME would do exactly what I needed to, but on Fabric it doesn't seem to be the case. And it doesn't seem to be the case with any other ActionResult.

Is this a Fabric bug? Or should I do something different?

Code of the listener:
```
ActionResult actionResult = ActionResult.PASS;
if (!player.isSpectator()) {
  BlockPos blockPos = result.getBlockPos();
  BlockState blockState = world.getBlockState(blockPos);
  Hand interactionHand = getInteractionHand(player);
  if(isCrop(blockState.getBlock()) && interactionHand != null) {
   try {
     IntProperty age = getAge(blockState);
     if (blockState.getOrEmpty(age).orElse(0) >= Collections.max(age.getValues())) {
       actionResult = ActionResult.CONSUME;
       if (!world.isClient()) {
          dropResources(world.getServer().getWorld(world.getRegistryKey()), blockState, player, blockPos, interactionHand, result.getSide());
          world.setBlockState(blockPos, blockState.with(age, Integer.valueOf(0)));
          player.swingHand(interactionHand, true);
        }
      }
    } catch (NullPointerException | NoSuchElementException | ClassCastException e) {
      e.printStackTrace();
    }
  }
}
return actionResult;
```

**Comment by warjort** [comment]
According to the javadoc, that event doesn't support CONSUME, only SUCCESS, PASS, FAIL
https://github.com/FabricMC/fabric/blob/1.18.2/fabric-events-interaction-v0/src/main/java/net/fabricmc/fabric/api/event/player/UseBlockCallback.java
but it also doesn't give an error if you return unsupported values.

From the mixin 
https://github.com/FabricMC/fabric/blob/640dbc8a5fc3ec1e2739ea6e1f8141a8f64b7a5b/fabric-events-interaction-v0/src/main/java/net/fabricmc/fabric/mixin/event/interaction/MixinClientPlayerInteractionManager.java#L97
you can see it will stop event handling if you return CONSUME but won't send a packet to the server which it does only for SUCCESS.

I would suggest you either submit a patch to make the event support CONSUME or write your own event/mixin that has the behaviour you want. 
Or you could just let the default player arm swing happen if you return SUCCESS.

**Reply by Juuxel** [reply]
Going from the API design POV, I'd say it should check for `ActionResult.isAccepted` instead of `== ActionResult.SUCCESS`, which matches what's done elsewhere in that mixin too. A potential PR, maybe?

**Reply by Technici4n** [reply]
IMO this is a badly designed legacy API that has generally unclear semantics. Further changes to it risk breaking mods that rely on these oddities.

**Reply by Crystal-Spider** [reply]
I would like to write my own event/mixin to handle correctly CONSUME, problem is I don't know what I need to do in order to handle CONSUME correctly.  
To be honest I still don't see why CONSUME is not handled correctly from the mixin, since CONSUME "indicates an action is performed but no animation should accompany the performance" (from ActionResult Javadoc), so it should be just like SUCCESS except no animation is played, which is exactly what it looks to me.  
Now, I'm quite new at all of this, so I'm evidently missing something, but I don't know how to proceed

**Comment by Crystal-Spider** [comment, marked as answer]
Solved it, I return `SUCCESS `instead of consume and instead of checking `interactionHand != null` I do `getInteractionHand(player) == hand`. This because I saw that the event is triggered 1 time per hand everytime, so I just had to wait for the correct hand.
